### ISSUE TITLE
Local Minimum

### ISSUE THREAD

**Issue by elisedl1** [rohitrango/FireANTs#54]
Hello,

I'm currently working with US-CT rigid registration with a custom loss function. For some cases, the registration works brilliantly, but for others it gets stuck at the very beginning at a local optima. I've tried tuning different Rigid Registration parameters, but nothing works so far. 

I know this is probably very problem specific, but I wanted to reach out and ask if there were any obvious solutions or ideas someone may have. These problem typically uses an Evolutionary Optimization approach, and I'm beginning to understand why..

Thanks!
Elise

**Comment by rohitrango** [comment]
Hi Elise,

Thanks for using FireANTs! I hope it helps in your research.

The quality of registration depends on the nature of your problem. If the relative orientation of your images are completely misaligned (anterior of US image is aligned to superior of CT image before registration), then gradient descent methods are unlikely to help, and you have to rely on a Bayesian optimization or evolutionary search algorithm. 
I'm actually building a search-based initialization for cases like histology to blockface where the relative orientations can be completely whack. It will take some time to validate though. cc @VedantZope

On the other hand, if the images are axis aligned, then it might be an issue with how rigid parameters are chosen. I'm working on making the rigid registration a little stable, you can try to optimize an affine transform directly though. 

If you can share more about the data and / or parameters you are using to optimize (like a jupyter notebook / python scripts) that will help me understand your problem better.

Cheers!

**Comment by elisedl1** [comment]
Hi!

Thanks for the reply!

I'm actually generating my own initial registration error directly from an aligned, ground-truth registration. I'm currently testing with a total initial TRE of around 12mm (papers of this paper will typically test anywhere from 1-40mm, with translation ranging from 0-20, and rotation between 0 and 5 degrees).

I'm able to get a final TRE of 1.15mm (great!), but if I try the same generated initial misalignment on a very similar new set of US-CT pairs, it will get stuck and I can only achieve a final TRE of 11mm, so its barely moving. 

Perhaps the trials that achieve good TREs are lucky and find a good optima by chance. Here are the parameters that work well:

rigid = RigidRegistration(
    [2,1],                     # pyramid_levels
    [1000,500],               # iterations_per_level
    batch1,                        # fixed_mask
    batch2,                       # moving_mask
    loss_type='custom',                # loss type
    custom_loss = custom_loss,
    optimizer='Adam',                  # optimizer type
    optimizer_lr=0.005,                 # learning rate
    cc_kernel_size=9,                   # cross-correlation kernel size
    blur=True
)

I've tried adding more pyramid levels, played around with iterations, and played around with the optimizer_lr. Even tweaking from 0.005 to 0.001 causes a terrible outcome. On the new set (similar intensities, distrbution, etc), I cannot get TRE down whatsoever. 

As well - for the nature of the problem, the moving image must remain rigid, as I'm registering rigid bone structures. It's 3D CT spine data to 3D US data. I'm registering one vertebra at a time, and some vertebra work great, but others I can't seem to escape local optima.
